drivers/sensor/st/lis3mdl/lis3mdl.c

```diff
@@ -163,55 +245,46 @@ int lis3mdl_init(const struct device *dev)
 	return 0;
 }
 
-#ifdef CONFIG_PM_DEVICE
-static int lis3mdl_pm_action(const struct device *dev, enum pm_device_action action)
-{
-	const struct lis3mdl_config *config = dev->config;
-	uint8_t ctrl_reg3;
-
-	switch (action) {
-	case PM_DEVICE_ACTION_RESUME:
-		ctrl_reg3 = LIS3MDL_MD_CONTINUOUS;
-
-		if (i2c_reg_write_byte_dt(&config->i2c, LIS3MDL_REG_CTRL3, ctrl_reg3) < 0) {
-			LOG_DBG("Failed to configure chip.");
-			return -EIO;
-		}
-
-		LOG_DBG("State changed to active");
-		break;
-	case PM_DEVICE_ACTION_SUSPEND:
-		ctrl_reg3 = LIS3MDL_MD_POWER_DOWN;
-
-		if (i2c_reg_write_byte_dt(&config->i2c, LIS3MDL_REG_CTRL3, ctrl_reg3) < 0) {
-			LOG_DBG("Failed to configure chip.");
-			return -EIO;
-		}
+/*
+ * Instantiation macros used when a device is on an SPI bus.
+ */
 
-		LOG_DBG("State changed to inactive");
-		break;
-	default:
-		return -ENOTSUP;
-	}
+#define LIS3MDL_SPI_OPERATION (SPI_WORD_SET(8) | SPI_OP_MODE_MASTER | SPI_MODE_CPOL | SPI_MODE_CPHA)
+
+#define LIS3MDL_CONFIG_SPI(inst)                                                                   \
+	static struct lis3mdl_data lis3mdl_data_##inst;                                            \
+                                                                                                   \
+	static struct lis3mdl_config lis3mdl_config_##inst = {                                     \
+		.spi = SPI_DT_SPEC_INST_GET(inst, LIS3MDL_SPI_OPERATION, 0),                       \
+		.write = lis3dml_write_spi,                                                        \
+		.read = lis3dml_read_spi,                                                          \
```

Comment:
This driver implementation is old. I think you should use STMEMSC_CTX_SPI and STMEMSC_CTX_I2C definitions, and get rid of the local read/write i2c/spi. Check with other ST drivers (like lis2dux12 for example)